doc/design/upgrade.md

```diff
@@ -125,15 +135,114 @@ Here is the upgrade workflow
+
+### Appliance log files
+
+Appliance log files configuration will not be touched at this time. So upgraded VCH will use old log files if there is any such kind of change. We can revisit this while there is customer requiement on this.
+
+### vSphere Object Management Logic
+
+There will have new vSphere API come up, so the logic to manage vsphere objects will be enhanced, for example, as Caglar said the vmdk can be managed directly in vSphere 6.5, instead of through vm operations in vSphere 6.0. We'll switch to these new interfaces for image and volume management sometime later.
+
+vic-machine is not supposed to migrate old image data or volume data to new vmdk files, so portlayer will need to be backward compatible. (Suppose vSphere will be backward compatibile, there should be nothing to do in vic, but need to do some research on it)
+
+### How to Support Upgrade Between Builds
+
+We need to keep old version's guestinfo structure definition, to make sure we can read back old configuration, and need plugin to migrate data from old structure to new structure. With current version mechanism, developer do not know which build version will be generated for his change, so we introduced another version, data migration version.
+
+If the value is not set, by default it is 0.
+
+Data migration framework will detect version difference between old version and latest and then run corresponding data migration plugins sequentially to migrate data.
+
+Notes:
+- Developers who change guestinfo, keyvalue store will be the owner to develop migration plugin, and be responsible to increase data migration version. 
+- Each migration version should have one and only one corresponding plugin.
+- If appliance configuration and container configuration are changed at the same time, two different plugins should be added, and registered to different plugin category.
+- If both appliance configuration and keyvalue store are changed, two plugins are recommended as well.
+
+### Limitation
+This document does not include approach for tether communication change, that means from serial port to VMCI or something else. Need to think about how to support old version's container while that change is made.
+
 ## Restrictions
 
 User cannot run two upgrades for same VCH at the same time. 
 
-vic-machine will check if there is already another upgrade snapshot is created before it start to create snapshot. But as create vsphere snapshot will take some time, e.g. one minute, if at this time, another upgrade process is started, it will start upgrade again cause the snapshot of previous task is not finished yet.
- 
\ No newline at end of file
+vic-machine will check if there is already another upgrade snapshot is created before it starts to create snapshot. But as create vsphere snapshot will take some time, e.g. one minute, if at this time, another upgrade process is started, it will start upgrade again cause the snapshot of previous task is not finished yet.
```

Comment:
It's also of use if the vSphere connection drops for some reason. Agree that detecting ISO files isn't a good option - I brought this up because, depending on the mechanism, there's a possibility to mis-identify a partial upgrade as in-progress which would stall the ability to retry.
This is the same scenario as `resume` if resume is implemented as revert/retry.